File: win-prog/ProgWinCS/Edit List and Spin/NotepadCloneWithRegistry/NotepadCloneWithRegistry.cs
//-------------------------------------------------------
// NotepadCloneWithRegistry.cs © 2001 by Charles Petzold
//-------------------------------------------------------
using Microsoft.Win32;
using System;
using System.Drawing;
using System.Windows.Forms;

class NotepadCloneWithRegistry: NotepadCloneNoMenu
{
     Rectangle        rectNormal;
     protected string strProgName;
     string           strRegKey    = "Software\\ProgrammingWindowsWithCSharp\\";
     const string     strWinState  = "WindowState";
     const string     strLocationX = "LocationX";
     const string     strLocationY = "LocationY";
     const string     strWidth     = "Width";
     const string     strHeight    = "Height";

     public new static void Main()
     {
          Application.Run(new NotepadCloneWithRegistry());
     }
     public NotepadCloneWithRegistry()
     {
          Text = strProgName = "Notepad Clone with Registry";
          rectNormal = DesktopBounds;
     }
     protected override void OnMove(EventArgs ea)
     {
          base.OnMove(ea);

          if (WindowState == FormWindowState.Normal)
               rectNormal = DesktopBounds;
     }
     protected override void OnResize(EventArgs ea)
     {
          base.OnResize(ea);

          if (WindowState == FormWindowState.Normal)
               rectNormal = DesktopBounds;
     }
     protected override void OnLoad(EventArgs ea)
     {
          base.OnLoad(ea);

               // Construct complete registry key.

          strRegKey = strRegKey + strProgName;

               // Load registry information.

          RegistryKey regkey = Registry.CurrentUser.OpenSubKey(strRegKey);

          if (regkey != null)
          {
               LoadRegistryInfo(regkey);
               regkey.Close();
          }
     }
     protected override void OnClosed(EventArgs ea)
     {
          base.OnClosed(ea);

               // Save registry information.

          RegistryKey regkey = 
                         Registry.CurrentUser.OpenSubKey(strRegKey, true);

          if (regkey == null)
               regkey = Registry.CurrentUser.CreateSubKey(strRegKey);

          SaveRegistryInfo(regkey);
          regkey.Close();
     }
     protected virtual void SaveRegistryInfo(RegistryKey regkey)
     {
          regkey.SetValue(strWinState,  (int) WindowState);
          regkey.SetValue(strLocationX, rectNormal.X);
          regkey.SetValue(strLocationY, rectNormal.Y);
          regkey.SetValue(strWidth,     rectNormal.Width);
          regkey.SetValue(strHeight,    rectNormal.Height);
     }
     protected virtual void LoadRegistryInfo(RegistryKey regkey)
     {
          int x  = (int) regkey.GetValue(strLocationX, 100);
          int y  = (int) regkey.GetValue(strLocationY, 100);
          int cx = (int) regkey.GetValue(strWidth, 300);
          int cy = (int) regkey.GetValue(strHeight, 300);

          rectNormal = new Rectangle(x, y, cx, cy);

               // Adjust rectangle for any change in desktop size.
          
          Rectangle rectDesk = SystemInformation.WorkingArea;

          rectNormal.Width  = Math.Min(rectNormal.Width,  rectDesk.Width);
          rectNormal.Height = Math.Min(rectNormal.Height, rectDesk.Height);
          rectNormal.X -= Math.Max(rectNormal.Right  - rectDesk.Right,  0);
          rectNormal.Y -= Math.Max(rectNormal.Bottom - rectDesk.Bottom, 0);

               // Set form properties.

          DesktopBounds = rectNormal;
          WindowState = (FormWindowState) regkey.GetValue(strWinState, 0);
     }
}

File: win-prog/ProgWinVB/Edit List and Spin/NotepadCloneWithRegistry/NotepadCloneWithRegistry.vb
'---------------------------------------------------------
' NotepadCloneWithRegistry.vb (c) 2002 by Charles Petzold
'---------------------------------------------------------
Imports Microsoft.Win32
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class NotepadCloneWithRegistry
    Inherits NotepadCloneNoMenu

    Protected strProgName As String
    Private rectNormal As Rectangle
    Private strRegKey As String = "Software\ProgrammingWindowsWithVBdotNet\"
    Const strWinState As String = "WindowState"
    Const strLocationX As String = "LocationX"
    Const strLocationY As String = "LocationY"
    Const strWidth As String = "Width"
    Const strHeight As String = "Height"

    Shared Shadows Sub Main()
        Application.Run(New NotepadCloneWithRegistry())
    End Sub

    Sub New()
        strProgName = "Notepad Clone with Registry"
        Text = strProgName
        rectNormal = DesktopBounds
    End Sub

    Protected Overrides Sub OnMove(ByVal ea As EventArgs)
        MyBase.OnMove(ea)
        If WindowState = FormWindowState.Normal Then
            rectNormal = DesktopBounds
        End If
    End Sub

    Protected Overrides Sub OnResize(ByVal ea As EventArgs)
        MyBase.OnResize(ea)
        If WindowState = FormWindowState.Normal Then
            rectNormal = DesktopBounds
        End If
    End Sub

    Protected Overrides Sub OnLoad(ByVal ea As EventArgs)
        MyBase.OnLoad(ea)

        ' Construct complete registry key.
        strRegKey = strRegKey & strProgName

        ' Load registry information.
        Dim regkey As RegistryKey = _
                        Registry.CurrentUser.OpenSubKey(strRegKey)
        If Not regkey Is Nothing Then
            LoadRegistryInfo(regkey)
            regkey.Close()
        End If
    End Sub

    Protected Overrides Sub OnClosed(ByVal ea As EventArgs)
        MyBase.OnClosed(ea)

        ' Save registry information.
        Dim regkey As RegistryKey = _
                        Registry.CurrentUser.OpenSubKey(strRegKey, True)
        If regkey Is Nothing Then
            regkey = Registry.CurrentUser.CreateSubKey(strRegKey)
        End If
        SaveRegistryInfo(regkey)
        regkey.Close()
    End Sub

    Protected Overridable Sub SaveRegistryInfo(ByVal regkey As RegistryKey)
        regkey.SetValue(strWinState, CInt(WindowState))
        regkey.SetValue(strLocationX, rectNormal.X)
        regkey.SetValue(strLocationY, rectNormal.Y)
        regkey.SetValue(strWidth, rectNormal.Width)
        regkey.SetValue(strHeight, rectNormal.Height)
    End Sub

    Protected Overridable Sub LoadRegistryInfo(ByVal regkey As RegistryKey)
        Dim x As Integer = DirectCast(regkey.GetValue(strLocationX, 100), _
                                      Integer)
        Dim y As Integer = DirectCast(regkey.GetValue(strLocationY, 100), _
                                      Integer)
        Dim cx As Integer = DirectCast(regkey.GetValue(strWidth, 300), _
                                       Integer)
        Dim cy As Integer = DirectCast(regkey.GetValue(strHeight, 300), _
                                       Integer)
        rectNormal = New Rectangle(x, y, cx, cy)

        ' Adjust rectangle for any change in desktop size.
        Dim rectDesk As Rectangle = SystemInformation.WorkingArea
        rectNormal.Width = Math.Min(rectNormal.Width, rectDesk.Width)
        rectNormal.Height = Math.Min(rectNormal.Height, rectDesk.Height)
        rectNormal.X -= Math.Max(rectNormal.Right - rectDesk.Right, 0)
        rectNormal.Y -= Math.Max(rectNormal.Bottom - rectDesk.Bottom, 0)

        ' Set form properties.
        DesktopBounds = rectNormal
        WindowState = CType(DirectCast(regkey.GetValue(strWinState, 0), _
                                       Integer), _
                            FormWindowState)
    End Sub
End Class
